pkg/ccl/backupccl/backup_span_maker.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package backupccl
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/rangedesc"
+	"github.com/cockroachdb/errors"
+)
+
+type requestSpanMaker struct {
+	requestSpans     []spanAndTime
+	specs            *execinfrapb.BackupDataSpec
+	rangeSizedSpans  bool
+	rangeIterFactory rangedesc.IteratorFactory
+}
+
+const maxChunkSize = 100
+
+// Aim to make at least 4 chunks per worker, ensuring size is >=1 and <= max.
+const softMinPerWorkerChunkCount = 4
+
+func (rsm *requestSpanMaker) queueWork(numSenders int) chan []spanAndTime {
+	todo := make(chan []spanAndTime, len(rsm.requestSpans))
+
+	chunkSize := (len(rsm.requestSpans) / (numSenders * softMinPerWorkerChunkCount)) + 1
+	if chunkSize > maxChunkSize {
+		chunkSize = maxChunkSize
+	}
+
+	chunk := make([]spanAndTime, 0, chunkSize)
+	for i := range rsm.requestSpans {
+		if !rsm.rangeSizedSpans {
+			todo <- []spanAndTime{rsm.requestSpans[i]}
+			continue
+		}
+
+		chunk = append(chunk, rsm.requestSpans[i])
+		if len(chunk) > chunkSize {
+			todo <- chunk
+			chunk = make([]spanAndTime, 0, chunkSize)
+		}
+	}
+	if len(chunk) > 0 {
+		todo <- chunk
+	}
+	return todo
+}
+
+func (rsm *requestSpanMaker) spanCountFromSpecs() int {
+	return len(rsm.specs.Spans) + len(rsm.specs.IntroducedSpans)
+}
+
+func initRequestSpanMaker(
+	rangeSizedSpans bool,
+	specs *execinfrapb.BackupDataSpec,
+	rangeIterfactory rangedesc.IteratorFactory,
+) *requestSpanMaker {
+	totalSpans := len(specs.Spans) + len(specs.IntroducedSpans)
+	return &requestSpanMaker{
+		requestSpans:     make([]spanAndTime, 0, totalSpans),
+		rangeSizedSpans:  rangeSizedSpans,
+		specs:            specs,
+		rangeIterFactory: rangeIterfactory,
+	}
+}
+
+func (rsm *requestSpanMaker) createRequestSpans(ctx context.Context) error {
+	if err := rsm.addSpans(ctx, rsm.specs.IntroducedSpans, hlc.Timestamp{}, rsm.specs.BackupStartTime); err != nil {
+		return err
+	}
+	if err := rsm.addSpans(ctx, rsm.specs.Spans, rsm.specs.BackupStartTime, rsm.specs.BackupEndTime); err != nil {
+		return err
+	}
+	log.Infof(ctx, "backup processor is assigned %d spans", rsm.spanCountFromSpecs())
```

Comment:
This logging feels like it belongs in the proc calling the helper rather than the helper

Reply:
i changed the log line to include the length of the request spans, which we don't expose to the proc.